ownCloud/Client/Actions/Actions+Extensions/DuplicateAction.swift

```diff
@@ -34,37 +37,126 @@ class DuplicateAction : Action {
 
 	// MARK: - Action implementation
 	override func run() {
-		guard context.items.count > 0, let core = self.core, let item = context.items.first, let itemName = item.name else {
+		guard context.items.count > 0 else {
 			completed(with: NSError(ocError: .itemNotFound))
 			return
 		}
+		remainingItems.append(contentsOf: context.items)
+		proceed()
+	}
 
-		context.items.forEach({ (item) in
-			if let rootItem = item.parentItem(from: core) {
+	func proceed() {
+		guard let item = remainingItems.first, let baseName = item.baseName, let core = self.core, let itemPath = item.path, let rootItemPath = itemPath.parentPath() else {
+			completed(with: NSError(ocError: .itemNotFound))
+			return
+		}
 
-				var name: String = "\(itemName) copy"
+		let pattern = " \(localizedCopy)[ ]?[0-9]{0,}$"
+		let searchName = baseName.replacingOccurrences(for: pattern)
 
-				if let itemFileExtension = item.fileExtension, let baseName = item.baseName {
-					var fileExtension = itemFileExtension
+		query = OCQuery(forPath: rootItemPath)
+		query?.delegate = self
+		query?.queryItem = item
 
-					if fileExtension != "" {
-						fileExtension = ".\(fileExtension)"
-					}
+		let filterHandler: OCQueryFilterHandler = { (_, _, item) -> Bool in
+			if let itemName = item?.baseName {
+				return itemName.hasPrefix(searchName)
+			}
+			return false
+		}
+		query?.addFilter(OCQueryFilter.init(handler: filterHandler), withIdentifier: "rename-file")
 
-					name = "\(baseName) copy\(fileExtension)"
+		core.start(query!)
+	}
+
+	// MARK: - Query handling
+	func query(_ query: OCQuery, failedWithError error: Error) {
+		completed(with: NSError(ocError: .itemNotFound))
+	}
+
+	func queryHasChangesAvailable(_ query: OCQuery) {
+		query.requestChangeSet(withFlags: OCQueryChangeSetRequestFlag(rawValue: 0)) { (query, changeSet) in
+
+			guard let queryItem = query.queryItem else { return }
+
+			switch query.state {
+			case .contentsFromCache, .waitingForServerReply:
+				break
+			case .idle:
+				self.core?.stop(query)
+
+				let items = changeSet?.queryResult ?? []
+				guard items.first != nil else {
+					OnBackgroundQueue {
+						self.copy(item: queryItem, with: 0)
+					}
+					return
 				}
 
-				if let progress = core.copy(item, to: rootItem, withName: name, options: nil, resultHandler: { (error, _, item, _) in
-					if error != nil {
-						Log.log("Error \(String(describing: error)) duplicating \(String(describing: item?.path))")
-						self.completed(with: error)
-					} else {
-						self.completed()
+				var fileCounter = -1
+				for anItem in items {
+					guard let baseName = anItem.baseName else { return }
+
+					let pattern = "\(self.localizedCopy)[ ]?[0-9]{0,}$"
```

Comment:
@hosy this pattern seems to be repeated, candidate for class-wide constant?

Reply:
It differs with a leading space

Reply:
@hosy ok